ipykernel/iostream.py

```diff
@@ -446,13 +447,18 @@ def __init__(
         self.pub_thread = pub_thread
         self.name = name
         self.topic = b"stream." + name.encode()
-        self.parent_header = {}
+        self._parent_header: contextvars.ContextVar[Dict[str, Any]] = contextvars.ContextVar(
+            "parent_header"
+        )
+        self._parent_header.set({})
+        self._thread_parents = {}
+        self._parent_header_global = {}
```

Comment:
Curious why the ContextVar and the _thread_parents dict? Is a threading.local not more standard?

Reply:
Because contextvar works well for asyncio edge cases. There is some more explanation in the PEP 567, but the gist is:

> Thread-local variables are insufficient for asynchronous tasks that execute concurrently in the same OS thread. Any context manager that saves and restores a context value using `threading.local()` will have its context values bleed to other code unexpectedly when used in async/await code.

That said, I will take another look at using `threading.local` instead of `_thread_parents`.

Reply:
I understand that we need ContextVar now, because also when we create new tasks we want to output to the right output cell, didn't think of that!

I think if we use ContextVar, we do not need the thread local storage, it's a superset of threading.local. In combination with overriding the Thread ctor, we can drop _thread_parents.